File: Strand/Data/StrengthPlanStore.swift
import Foundation

/// User-authored strength workouts (sets / reps / muscle groups).
/// Twin of Android `StrengthPlanStore` — slim port for iOS parity (8.6.164).
/// Local-only; never invents HR or cardio Effort. Volume is note-based until a session is logged.
final class StrengthPlanStore: ObservableObject {
    static let shared = StrengthPlanStore()

    private let defaults = UserDefaults.standard
    private let key = "noop_strength_plans_json"

    @Published private(set) var plans: [Plan] = []

    struct Exercise: Identifiable, Equatable, Codable {
        var id: Int64
        var name: String
        var muscleGroup: String
        var sets: Int
        var reps: Int
        var weightKg: Double?
        var notes: String

        var volumeKg: Double {
            (weightKg ?? 0) * Double(max(sets, 0)) * Double(max(reps, 0))
        }

        func summaryLine() -> String {
            let w = weightKg.map { " · \(Self.formatKg($0)) kg" } ?? ""
            return "\(name) · \(sets)×\(reps)\(w) · \(muscleGroup)"
        }

        static func formatKg(_ v: Double) -> String {
            if v >= 100 { return "\(Int(v.rounded()))" }
            var s = String(format: "%.1f", v)
            if s.hasSuffix(".0") { s = String(s.dropLast(2)) }
            return s
        }
    }

    struct Plan: Identifiable, Equatable, Codable {
        var id: Int64
        var name: String
        var exercises: [Exercise]
        var createdAtMs: Int64
        var updatedAtMs: Int64
        var lastUsedAtMs: Int64
        var notes: String

        var totalSets: Int { exercises.reduce(0) { $0 + max($1.sets, 0) } }
        var totalMoves: Int { exercises.count }
        var estimatedVolumeKg: Double { exercises.reduce(0) { $0 + $1.volumeKg } }

        func sessionNotes() -> String {
            let lines = exercises.map { e -> String in
                let w = e.weightKg.map { " @\(Exercise.formatKg($0))kg" } ?? ""
                return "\(e.name) \(e.sets)×\(e.reps)\(w)"
            }.joined(separator: " · ")
            let vol = estimatedVolumeKg > 0
                ? " · volume load \(Exercise.formatKg(estimatedVolumeKg)) kg"
                : ""
            let body = "Strength plan \"\(name)\" · \(exercises.count) exercises · \(totalSets) sets\(vol) · \(lines)"
            return String(body.prefix(480))
        }
    }

    static let muscleGroups = [
        "chest", "back", "shoulders", "arms", "core",
        "quads", "hamstrings", "glutes", "calves", "full",
    ]

    static let planNamePresets = ["Push day", "Pull day", "Legs", "Full body", "Upper", "Lower"]

    static let exercisePresets: [String: [String]] = [
        "chest": ["Bench press", "Incline press", "Chest fly", "Push-up"],
        "back": ["Barbell row", "Lat pulldown", "Pull-up", "Seated row"],
        "shoulders": ["OHP", "Lateral raise", "Face pull", "Arnold press"],
        "arms": ["Curl", "Tricep pushdown", "Hammer curl", "Skull crusher"],
        "core": ["Plank", "Cable crunch", "Hanging knee raise", "Dead bug"],
        "quads": ["Back squat", "Leg press", "Lunge", "Leg extension"],
        "hamstrings": ["RDL", "Leg curl", "Good morning", "Nordic curl"],
        "glutes": ["Hip thrust", "Glute bridge", "Cable kickback", "Step-up"],
        "calves": ["Standing calf raise", "Seated calf raise"],
        "full": ["Deadlift", "Clean", "Thruster", "Farmer carry"],
    ]

    private init() { reload() }

    func reload() {
        guard let data = defaults.data(forKey: key),
              let decoded = try? JSONDecoder().decode([Plan].self, from: data) else {
            plans = []
            return
        }
        plans = decoded.sorted {
            max($0.lastUsedAtMs, $0.updatedAtMs) > max($1.lastUsedAtMs, $1.updatedAtMs)
        }
    }

    private func persist() {
        let capped = Array(plans.prefix(40))
        plans = capped
        if let data = try? JSONEncoder().encode(capped) {
            defaults.set(data, forKey: key)
        }
    }

    func upsert(_ plan: Plan) {
        var all = plans
        let stamped = Plan(
            id: plan.id,
            name: plan.name,
            exercises: plan.exercises,
            createdAtMs: plan.createdAtMs,
            updatedAtMs: Int64(Date().timeIntervalSince1970 * 1000),
            lastUsedAtMs: plan.lastUsedAtMs,
            notes: plan.notes
        )
        if let idx = all.firstIndex(where: { $0.id == stamped.id }) {
            all[idx] = stamped
        } else {
            all.append(stamped)
        }
        plans = all
        persist()
        reload()
    }

    func delete(id: Int64) {
        plans = plans.filter { $0.id != id }
        persist()
    }

    @discardableResult
    func duplicate(id: Int64) -> Plan? {
        guard let src = plans.first(where: { $0.id == id }) else { return nil }
        let now = Int64(Date().timeIntervalSince1970 * 1000)
        let copy = Plan(
            id: now,
            name: String((src.name + " copy").prefix(60)),
            exercises: src.exercises.enumerated().map { i, e in
                var e2 = e
                e2.id = now + 1 + Int64(i)
                return e2
            },
            createdAtMs: now,
            updatedAtMs: now,
            lastUsedAtMs: 0,
            notes: src.notes
        )
        upsert(copy)
        return copy
    }

    func markUsed(id: Int64) {
        guard let idx = plans.firstIndex(where: { $0.id == id }) else { return }
        var p = plans[idx]
        let now = Int64(Date().timeIntervalSince1970 * 1000)
        p.lastUsedAtMs = now
        p.updatedAtMs = now
        plans[idx] = p
        persist()
        reload()
    }

    static func normalizeMuscle(_ raw: String) -> String {
        let t = raw.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        if muscleGroups.contains(t) { return t }
        if t.hasPrefix("ham") { return "hamstrings" }
        if t == "legs" || t == "leg" { return "quads" }
        if t == "abs" || t == "ab" { return "core" }
        if t == "delt" || t == "delts" { return "shoulders" }
        if ["bicep", "tricep", "biceps", "triceps"].contains(t) { return "arms" }
        return "full"
    }

    static func starterPlan(name: String) -> Plan {
        let now = Int64(Date().timeIntervalSince1970 * 1000)
        let muscle: String = {
            switch name.lowercased() {
            case "push day", "upper": return "chest"
            case "pull day": return "back"
            case "legs", "lower": return "quads"
            default: return "full"
            }
        }()
        let names = (exercisePresets[muscle] ?? []).prefix(3)
        let list = names.isEmpty ? ["Compound lift"] : Array(names)
        return Plan(
            id: now,
            name: name,
            exercises: list.enumerated().map { i, n in
                Exercise(
                    id: now + 1 + Int64(i),
                    name: n,
                    muscleGroup: muscle,
                    sets: 3,
                    reps: muscle == "core" ? 12 : 8,
                    weightKg: nil,
                    notes: ""
                )
            },
            createdAtMs: now,
            updatedAtMs: now,
            lastUsedAtMs: 0,
            notes: ""
        )
    }
}

File: android/app/src/main/java/com/noop/data/StrengthPlanStore.kt
package com.noop.data

import android.content.Context
import org.json.JSONArray
import org.json.JSONObject

/**
 * User-authored strength workouts (sets / reps / muscle groups).
 * Local-only; never invents HR or cardio Effort. Volume is note-based until a session is logged.
 */
class StrengthPlanStore private constructor(context: Context) {
    private val prefs = context.applicationContext.getSharedPreferences(PREFS, Context.MODE_PRIVATE)

    data class Exercise(
        val id: Long,
        val name: String,
        val muscleGroup: String,
        val sets: Int,
        val reps: Int,
        val weightKg: Double? = null,
        val notes: String = "",
    ) {
        val volumeKg: Double
            get() = (weightKg ?: 0.0) * sets.coerceAtLeast(0) * reps.coerceAtLeast(0)

        fun summaryLine(): String {
            val w = weightKg?.let { " · ${formatKg(it)} kg" }.orEmpty()
            return "$name · $sets×$reps$w · $muscleGroup"
        }
    }

    data class Plan(
        val id: Long,
        val name: String,
        val exercises: List<Exercise>,
        val createdAtMs: Long = System.currentTimeMillis(),
        val updatedAtMs: Long = System.currentTimeMillis(),
        val lastUsedAtMs: Long = 0L,
        val notes: String = "",
    ) {
        val totalSets: Int get() = exercises.sumOf { it.sets.coerceAtLeast(0) }
        val totalMoves: Int get() = exercises.size
        val estimatedVolumeKg: Double get() = exercises.sumOf { it.volumeKg }
        val primaryMuscle: String
            get() = exercises.groupingBy { it.muscleGroup }
                .eachCount()
                .maxByOrNull { it.value }
                ?.key
                ?: "full"

        fun sessionNotes(): String {
            val lines = exercises.joinToString(" · ") {
                val w = it.weightKg?.let { kg -> " @${formatKg(kg)}kg" }.orEmpty()
                "${it.name} ${it.sets}×${it.reps}$w"
            }
            val vol = estimatedVolumeKg.takeIf { it > 0 }?.let { " · volume load ${formatKg(it)} kg" }.orEmpty()
            return "Strength plan \"$name\" · ${exercises.size} exercises · $totalSets sets$vol · $lines"
                .take(480)
        }
    }

    fun loadPlans(): List<Plan> {
        val raw = prefs.getString(KEY, "[]") ?: "[]"
        return runCatching {
            val arr = JSONArray(raw)
            buildList {
                for (i in 0 until arr.length()) {
                    val o = arr.getJSONObject(i)
                    val ex = o.optJSONArray("exercises") ?: JSONArray()
                    val exercises = buildList {
                        for (j in 0 until ex.length()) {
                            val e = ex.getJSONObject(j)
                            add(
                                Exercise(
                                    id = e.optLong("id"),
                                    name = e.optString("name").ifBlank { "Exercise" },
                                    muscleGroup = normalizeMuscle(e.optString("muscleGroup", "full")),
                                    sets = e.optInt("sets", 3).coerceIn(1, 20),
                                    reps = e.optInt("reps", 8).coerceIn(1, 50),
                                    weightKg = e.optDouble("weightKg").takeIf { !it.isNaN() && it > 0 },
                                    notes = e.optString("notes"),
                                ),
                            )
                        }
                    }
                    add(
                        Plan(
                            id = o.optLong("id"),
                            name = o.optString("name").ifBlank { "Workout" },
                            exercises = exercises,
                            createdAtMs = o.optLong("createdAtMs"),
                            updatedAtMs = o.optLong("updatedAtMs"),
                            lastUsedAtMs = o.optLong("lastUsedAtMs"),
                            notes = o.optString("notes"),
                        ),
                    )
                }
            }.sortedWith(
                compareByDescending<Plan> { it.lastUsedAtMs.coerceAtLeast(it.updatedAtMs) }
                    .thenByDescending { it.updatedAtMs },
            )
        }.getOrDefault(emptyList())
    }

    fun savePlans(plans: List<Plan>) {
        val arr = JSONArray()
        for (p in plans.takeLast(40)) {
            val ex = JSONArray()
            for (e in p.exercises.take(40)) {
                ex.put(
                    JSONObject()
                        .put("id", e.id)
                        .put("name", e.name.trim().take(80))
                        .put("muscleGroup", normalizeMuscle(e.muscleGroup))
                        .put("sets", e.sets.coerceIn(1, 20))
                        .put("reps", e.reps.coerceIn(1, 50))
                        .put("weightKg", e.weightKg ?: JSONObject.NULL)
                        .put("notes", e.notes.take(120)),
                )
            }
            arr.put(
                JSONObject()
                    .put("id", p.id)
                    .put("name", p.name.trim().take(60))
                    .put("exercises", ex)
                    .put("createdAtMs", p.createdAtMs)
                    .put("updatedAtMs", p.updatedAtMs)
                    .put("lastUsedAtMs", p.lastUsedAtMs)
                    .put("notes", p.notes.take(200)),
            )
        }
        prefs.edit().putString(KEY, arr.toString()).apply()
    }

    fun upsert(plan: Plan) {
        val all = loadPlans().toMutableList()
        val idx = all.indexOfFirst { it.id == plan.id }
        val stamped = plan.copy(updatedAtMs = System.currentTimeMillis())
        if (idx >= 0) all[idx] = stamped else all.add(stamped)
        savePlans(all)
    }

    fun delete(id: Long) {
        savePlans(loadPlans().filterNot { it.id == id })
    }

    fun duplicate(id: Long): Plan? {
        val src = loadPlans().firstOrNull { it.id == id } ?: return null
        val now = System.currentTimeMillis()
        val copy = src.copy(
            id = now,
            name = "${src.name} copy".take(60),
            createdAtMs = now,
            updatedAtMs = now,
            lastUsedAtMs = 0L,
            exercises = src.exercises.mapIndexed { i, e -> e.copy(id = now + 1 + i) },
        )
        upsert(copy)
        return copy
    }

    fun markUsed(id: Long) {
        val all = loadPlans().toMutableList()
        val idx = all.indexOfFirst { it.id == id }
        if (idx < 0) return
        all[idx] = all[idx].copy(lastUsedAtMs = System.currentTimeMillis(), updatedAtMs = System.currentTimeMillis())
        savePlans(all)
    }

    companion object {
        private const val PREFS = "noop_strength_plans"
        private const val KEY = "plans_json"

        val MUSCLE_GROUPS = listOf(
            "chest", "back", "shoulders", "arms", "core",
            "quads", "hamstrings", "glutes", "calves", "full",
        )

        val PLAN_NAME_PRESETS = listOf("Push day", "Pull day", "Legs", "Full body", "Upper", "Lower")

        val EXERCISE_PRESETS: Map<String, List<String>> = mapOf(
            "chest" to listOf("Bench press", "Incline press", "Chest fly", "Push-up"),
            "back" to listOf("Barbell row", "Lat pulldown", "Pull-up", "Seated row"),
            "shoulders" to listOf("OHP", "Lateral raise", "Face pull", "Arnold press"),
            "arms" to listOf("Curl", "Tricep pushdown", "Hammer curl", "Skull crusher"),
            "core" to listOf("Plank", "Cable crunch", "Hanging knee raise", "Dead bug"),
            "quads" to listOf("Back squat", "Leg press", "Lunge", "Leg extension"),
            "hamstrings" to listOf("RDL", "Leg curl", "Good morning", "Nordic curl"),
            "glutes" to listOf("Hip thrust", "Glute bridge", "Cable kickback", "Step-up"),
            "calves" to listOf("Standing calf raise", "Seated calf raise"),
            "full" to listOf("Deadlift", "Clean", "Thruster", "Farmer carry"),
        )

        fun normalizeMuscle(raw: String): String {
            val t = raw.trim().lowercase()
            return when {
                t in MUSCLE_GROUPS -> t
                t.startsWith("ham") -> "hamstrings"
                t == "legs" || t == "leg" -> "quads"
                t == "abs" || t == "ab" -> "core"
                t == "delt" || t == "delts" -> "shoulders"
                t == "bicep" || t == "tricep" || t == "biceps" || t == "triceps" -> "arms"
                else -> "full"
            }
        }

        fun starterPlan(name: String): Plan {
            val now = System.currentTimeMillis()
            val muscle = when (name.lowercase()) {
                "push day", "upper" -> "chest"
                "pull day" -> "back"
                "legs", "lower" -> "quads"
                else -> "full"
            }
            val names = EXERCISE_PRESETS[muscle].orEmpty().take(3).ifEmpty { listOf("Compound lift") }
            return Plan(
                id = now,
                name = name,
                exercises = names.mapIndexed { i, n ->
                    Exercise(
                        id = now + 1 + i,
                        name = n,
                        muscleGroup = muscle,
                        sets = 3,
                        reps = if (muscle == "core") 12 else 8,
                    )
                },
                createdAtMs = now,
                updatedAtMs = now,
            )
        }

        fun formatKg(v: Double): String =
            if (v >= 100) v.toInt().toString() else String.format(java.util.Locale.US, "%.1f", v).trimEnd('0').trimEnd('.')

        @Volatile private var instance: StrengthPlanStore? = null
        fun from(context: Context): StrengthPlanStore {
            val app = context.applicationContext
            return instance ?: synchronized(this) {
                instance ?: StrengthPlanStore(app).also { instance = it }
            }
        }
    }
}
